File: services/exploreService.js
const characters = require('../utils/data/characters');
const db = require('../database');

/**
 * Explore Service: Manages 1v1 encounter generation and capture logic.
 */
class ExploreService {
    constructor() {
        this.locks = new Set();
        this.CONSTANTS = {
            STAMINA_COST: 0,
            MAX_STAMINA: 100,
            REFILL_MINUTES: 5,
            DAILY_EXPLORE_LIMIT: 1000,
            DAILY_CATCH_LIMIT: 10,
            COOLDOWN_SEC: 5,
            MAX_STEPS: 3,
            
            // Base Rates
            BASE_CHANCES: {
                "Common": 0.35,
                "Rare": 0.18,
                "Epic": 0.08,
                "Legendary": 0.03,
                "Mythic": 0.005
            },
// ... (REWARDS)

            // WIN REWARDS
            REWARDS: {
                "Common": { coins: 30, xp: 15 },
                "Rare": { coins: 60, xp: 25 },
                "Epic": { coins: 120, xp: 40 },
                "Legendary": { coins: 250, xp: 70 },
                "Mythic": { coins: 1000, xp: 200 }
            },
            LOSS_REWARDS: { coins: 20, xp: 5 }
        };

        this.BIOMES = {
            "TIER_1": {
                id: "TIER_1",
                name: "🏮 Haunted Outskirts",
                desc: "Low-level spirits (Lv.1+).",
                minLevel: 1,
                multiplier: 1.0,
                probabilities: { "Common": 0.84, "Rare": 0.10, "Epic": 0.05, "Legendary": 0.01, "Mythic": 0.0 }
            },
            "TIER_2": {
                id: "TIER_2",
                name: "🏚️ Cursed Urban District",
                desc: "Stronger curses (Lv.20+).",
                minLevel: 20,
                multiplier: 1.8,
                probabilities: { "Common": 0.35, "Rare": 0.50, "Epic": 0.12, "Legendary": 0.03, "Mythic": 0.0 }
            },
            "TIER_3": {
                id: "TIER_3",
                name: "🏯 Special Grade Territory",
                desc: "Special Grade threats (Lv.50+).",
                minLevel: 50,
                multiplier: 3.5,
                probabilities: { "Common": 0.0, "Rare": 0.20, "Epic": 0.70, "Legendary": 0.08, "Mythic": 0.02 }
            }
        };

        this.CONSTANTS.REWARDS = {
            "Common": { coins: 100, xp: 50, dust: 20, shards: 5 },
            "Rare": { coins: 300, xp: 120, dust: 50, shards: 15 },
            "Epic": { coins: 1000, xp: 400, dust: 150, shards: 50 },
            "Legendary": { coins: 5000, xp: 1500, dust: 500, shards: 200 },
            "Mythic": { coins: 20000, xp: 10000, dust: 2000, shards: 1000 }
        };
    }

    acquireLock(userId) {
        if (this.locks.has(userId)) return false;
        this.locks.add(userId);
        // Safety timeout to prevent permanent locks
        setTimeout(() => this.locks.delete(userId), 15000); 
        return true;
    }

    releaseLock(userId) {
        this.locks.delete(userId);
    }

    async rollEncounter(biomeKey = "ACADEMY", forceRarity = null) {
        const biome = this.BIOMES[biomeKey] || this.BIOMES["ACADEMY"];
        const roll = Math.random();
        let cumulative = 0;
        let selectedRarity = forceRarity || "Common";

        if (!forceRarity) {
            for (const [rarity, chance] of Object.entries(biome.probabilities)) {
                cumulative += chance;
                if (roll <= cumulative) {
                    selectedRarity = rarity;
                    break;
                }
            }
        }

        const pool = Object.values(characters).filter(c => c.rarity === selectedRarity);
        if (pool.length === 0) return this.rollEncounter("ACADEMY");

        const selected = pool[Math.floor(Math.random() * pool.length)];

        return {
            character: selected,
            rarity: selectedRarity,
            biome: biomeKey,
            catchable: this.CONSTANTS.BASE_CHANCES[selectedRarity] > 0
        };
    }

    getRandomEvent(currentStep) {
        const events = [
            { type: 'battle', weight: 60, icon: '⚔️', name: 'CURSED SPIRIT' },
            { type: 'scavenge', weight: 20, icon: '📦', name: 'TREASURE CACHE' },
            { type: 'mystery', weight: 10, icon: '❔', name: 'SHADOWY FIGURE' },
            { type: 'rest', weight: 10, icon: '⛺', name: 'SAFE ZONE' }
        ];

        // Boss on final step
        if (currentStep >= this.CONSTANTS.MAX_STEPS) return events[0];

        const roll = Math.random() * 100;
        let cumulative = 0;
        for (const e of events) {
            cumulative += e.weight;
            if (roll <= cumulative) return e;
        }
        return events[0];
    }

    generateStepReward(type) {
        const roll = Math.random();
        switch(type) {
            case 'scavenge': 
                const shardChance = Math.random() < 0.10;
                if (shardChance) return { shardsCurrency: 10, msg: "📦 <b>LUCKY FIND!</b>\nYou found a cache of Cursed Shards!\n🧩 +10 Shards" };
                return { coins: 250, dust: 10, msg: "📦 You found a supply crate!\n💰 +250 Coins\n✨ +10 Dust" };
            case 'treasure':
                const item = roll < 0.7 ? 'cursed_charm' : 'gacha_ticket';
                return { 
                    coins: 750, 
                    itemId: item, 
                    msg: `💎 <b>TREASURE CACHE!</b>\n💰 +750 Coins\n🎁 Found: 1x ${item.replace(/_/g, ' ').toUpperCase()}` 
                };
            case 'rest': 
                return { stamina: 50, msg: "🍵 A brief respite at a safe house.\n⚡ +50 Stamina restored." };
            case 'mystery': 
                if (roll < 0.4) return { shardsCurrency: 15, msg: "🌑 A shadow figure grants you power...\n🧩 +15 Shards found!" };
                return { coins: 200, xp: 150, msg: "📖 A wandering soul shares its knowledge.\n💰 +200 Coins\n📈 +150 XP" };
            default: return { coins: 20, msg: "💨 Just some wind through the ruins." };
        }
    }

    async calculateFinalCaptureChance(user, targetChar, battleData) {
        const base = this.CONSTANTS.BASE_CHANCES[targetChar.rarity] || 0.1;
        let finalChance = base;
        const modifiers = [];

        // 1. Capture Items
        if (battleData.usedItem === 'domain_essence') {
            return { finalChance: 1.0, modifiers: ["🌌 Domain Essence (100% Guaranteed)"] };
        }
        
        if (battleData.usedItem === 'grade_1_shackle') {
            finalChance += 0.30;
            modifiers.push("⛓ Grade-1 Shackle (+30%)");
        } else if (battleData.usedItem === 'cursed_seal_tag') {
            finalChance += 0.15;
            modifiers.push("🏷️ Cursed Seal (+15%)");
        } else if (battleData.usedItem === 'cursed_charm' || battleData.usedCharm || user.activeCursedCharm) {
            finalChance += 0.25;
            modifiers.push("🧿 Cursed Charm (+25%)");
        }

        // 2. HP Scaling
        const hpPct = (battleData.enemyHp / battleData.enemyMaxHp) * 100;
        if (hpPct < 15) {
            finalChance += 0.15;
            modifiers.push("🩸 Critical Health (+15%)");
        } else if (hpPct <= 40) {
            finalChance += 0.05;
            modifiers.push("🩹 Low Health (+5%)");
        }

        // 3. Damage Contribution (+10%)
        if (battleData.dmgP1Percent > 90) {
            finalChance += 0.10;
            modifiers.push("⚔️ Overwhelming Might (+10%)");
        }

        // 4. Black Flash Finisher (+10%)
        if (battleData.lastHitWasBlackFlash) {
            finalChance += 0.10;
            modifiers.push("⚡ Black Flash Finish (+10%)");
        }

        // 5. Duplicate Penalty (-30%)
        const alreadyOwns = await db.roster.findOne({ userId: user.telegramId, charId: targetChar.name });
        if (alreadyOwns) {
            finalChance -= 0.10;
            modifiers.push("🔄 Duplicate Bond (-10%)");
        } else {
            finalChance += 0.10;
            modifiers.push("✨ First Discovery (+10%)");
        }

        // 6. Level Gap Penalty (-10%)
        if (targetChar.level > (user.playerLevel || 1) + 10) {
            finalChance -= 0.10;
            modifiers.push("🏔 Level Gap (-10%)");
        }

        return {
            finalChance: Math.max(0.01, Math.min(0.80, finalChance)),
            modifiers
        };
    }

    syncStamina(user) {
        return null;
    }

    /**
     * Auto-Grind: Performs 10 automated hunts at once.
     * Rewards are standard, but character encounters are much rarer (30% of normal).
     * Automatically attempts a low-chance (5%) capture if a character is found.
     */
    async runAutoGrind(user, biomeKey = "TIER_1") {
        const batchSize = 10;
        const staminaCost = 50; // 5 per hunt
        const results = {
            coins: 0,
            dust: 0,
            xp: 0,
            charactersFound: [],
            shards: 0,
            staminaUsed: staminaCost
        };

        if ((user.stamina || 0) < staminaCost) {
            return { success: false, msg: `❌ Insufficient Stamina! You need 🔋 50 Stamina for this blitz.` };
        }

        const biome = this.BIOMES[biomeKey] || this.BIOMES["TIER_1"];
        
        for (let i = 0; i < batchSize; i++) {
            // 1. Base Rewards (Simulate 1 step of exploration)
            const event = this.getRandomEvent(1); // Usually a battle or scavenge
            
            if (event.type === 'battle') {
                // Reduced character chance (30% of normal)
                if (Math.random() < 0.30) {
                    const encounter = await this.rollEncounter(biomeKey);
                    
                    // Auto-Capture check (Flat 5% chance)
                    if (Math.random() < 0.05) {
                        results.charactersFound.push(encounter.character.name);
                    } else {
                        // Consolation prize for finding but not catching
                        results.coins += 50;
                    }
                } else {
                    // Standard trash mob reward
                    results.coins += 30;
                    results.xp += 15;
                }
            } else {
                const reward = this.generateStepReward(event.type);
                results.coins += (reward.coins || 0);
                results.dust += (reward.dust || 0);
                results.xp += (reward.xp || 0);
                results.shards += (reward.shardsCurrency || 0);
            }
        }

        return { success: true, ...results };
    }

    checkDailyReset(user) {
        if (!user) return null;
        const now = new Date();
        const lastReset = user.lastDailyReset ? new Date(user.lastDailyReset) : new Date(0);
        
        if (now.getUTCDate() !== lastReset.getUTCDate() || now.getUTCMonth() !== lastReset.getUTCMonth()) {
            return {
                dailyExploreCount: 0,
                dailyCatchCount: 0,
                lastDailyReset: now.getTime()
            };
        }
        return null;
    }
}

module.exports = new ExploreService();

File: services/explore_service.py
import random
import time
from datetime import datetime
from database import db
from utils.data import characters

class ExploreService:
    def __init__(self):
        self.locks = set()
        self.last_encountered = {} # user_id -> list of last 5 names
        self.CONSTANTS = {
            "STAMINA_COST": 0,
            "MAX_STAMINA": 100,
            "REFILL_MINUTES": 5,
            "DAILY_EXPLORE_LIMIT": 1000,
            "MAX_STEPS": 3,
            "BASE_CHANCES": {},
            "REWARDS": {
                "Common": {"coins": 100, "xp": 50, "dust": 20, "shards": 5},
                "Rare": {"coins": 300, "xp": 120, "dust": 50, "shards": 15},
                "Epic": {"coins": 1000, "xp": 400, "dust": 150, "shards": 50},
                "Legendary": {"coins": 5000, "xp": 1500, "dust": 500, "shards": 200},
                "Mythic": {"coins": 20000, "xp": 10000, "dust": 2000, "shards": 1000}
            },
            "LOSS_REWARDS": {"coins": 20, "xp": 5}
        }

        self.BIOMES = {
            "TIER_1": {
                "id": "TIER_1",
                "name": "🏮 Haunted Outskirts",
                "desc": "Low-level spirits (Lv.1+).",
                "minLevel": 1,
                "multiplier": 1.0,
                "probabilities": {"Common": 0.838, "Rare": 0.10, "Epic": 0.05, "Legendary": 0.01, "Mythic": 0.002}
            },
            "TIER_2": {
                "id": "TIER_2",
                "name": "🏚️ Cursed Urban District",
                "desc": "Stronger curses (Lv.20+).",
                "minLevel": 20,
                "multiplier": 1.8,
                "probabilities": {"Common": 0.345, "Rare": 0.50, "Epic": 0.12, "Legendary": 0.03, "Mythic": 0.005}
            },
            "TIER_3": {
                "id": "TIER_3",
                "name": "🏯 Special Grade Territory",
                "desc": "Special Grade threats (Lv.50+).",
                "minLevel": 50,
                "multiplier": 3.5,
                "probabilities": {"Common": 0.0, "Rare": 0.20, "Epic": 0.70, "Legendary": 0.08, "Mythic": 0.02}
            }
        }

    def acquire_lock(self, user_id):
        if user_id in self.locks:
            return False
        self.locks.add(user_id)
        # We'd ideally use an async sleep or timer to release lock, 
        # but for simplicity in Python we handle it in the handler
        return True

    def release_lock(self, user_id):
        if user_id in self.locks:
            self.locks.remove(user_id)

    async def roll_encounter(self, user_id=None, biome_key="TIER_1", force_rarity=None):
        biome = self.BIOMES.get(biome_key, self.BIOMES["TIER_1"])
        roll = random.random()
        cumulative = 0
        selected_rarity = force_rarity or "Common"

        if not force_rarity:
            for rarity, chance in biome["probabilities"].items():
                cumulative += chance
                if roll <= cumulative:
                    selected_rarity = rarity
                    break

        # 2. Select Character from Pool
        pool = [c for c in characters.DATA.values() if c.get('rarity') == selected_rarity]
        if not pool:
            # Fallback to Common if requested rarity pool is empty
            pool = [c for c in characters.DATA.values() if c.get('rarity') == 'Common']
        
        if not pool:
            # Absolute fallback if even common pool is empty (should not happen)
            return await self.roll_encounter(user_id, "TIER_1")

        # Avoid picking any of the last 5 characters for this user
        history = self.last_encountered.get(user_id, [])
        if len(pool) > len(history):
            temp_pool = [c for c in pool if c['name'] not in history]
            if temp_pool:
                pool = temp_pool

        # Shuffle pool for extra randomness
        random.shuffle(pool)
        selected = random.choice(pool)


        # Update history (keep last 5)
        history.append(selected['name'])
        if len(history) > 5:
            history.pop(0)
        self.last_encountered[user_id] = history

        return {
            "character": selected,
            "rarity": selected_rarity,
            "biome": biome_key,
            "catchable": True # All are catchable now
        }

    def get_random_event(self, current_step):
        events = [
            {"type": 'battle', "weight": 60, "icon": '⚔️', "name": 'CURSED SPIRIT'},
            {"type": 'scavenge', "weight": 20, "icon": '📦', "name": 'TREASURE CACHE'},
            {"type": 'mystery', "weight": 10, "icon": '❔', "name": 'SHADOWY FIGURE'},
            {"type": 'rest', "weight": 10, "icon": '⛺', "name": 'SAFE ZONE'}
        ]

        if current_step >= self.CONSTANTS["MAX_STEPS"]:
            return events[0]

        roll = random.random() * 100
        cumulative = 0
        for e in events:
            cumulative += e["weight"]
            if roll <= cumulative:
                return e
        return events[0]

    def generate_step_reward(self, event_type):
        roll = random.random()
        if event_type == 'scavenge':
            if random.random() < 0.10:
                return {"shardsCurrency": 10, "msg": "📦 <b>LUCKY FIND!</b>\nYou found a cache of Cursed Shards!\n🧩 +10 Shards"}
            return {"coins": 250, "dust": 10, "msg": "📦 You found a supply crate!\n💰 +250 Coins\n✨ +10 Dust"}
        elif event_type == 'treasure':
            item = 'cursed_charm' if roll < 0.7 else 'gacha_ticket'
            return {
                "coins": 750,
                "itemId": item,
                "msg": f"💎 <b>TREASURE CACHE!</b>\n💰 +750 Coins\n🎁 Found: 1x {item.replace('_', ' ').upper()}"
            }
        elif event_type == 'rest':
            return {"stamina": 50, "msg": "🍵 A brief respite at a safe house.\n⚡ +50 Stamina restored."}
        elif event_type == 'mystery':
            if roll < 0.4:
                return {"shardsCurrency": 15, "msg": "🌑 A shadow figure grants you power...\n🧩 +15 Shards found!"}
            return {"coins": 200, "xp": 150, "msg": "📖 A wandering soul shares its knowledge.\n💰 +200 Coins\n📈 +150 XP"}
        else:
            return {"coins": 20, "msg": "💨 Just some wind through the ruins."}

    async def run_auto_grind(self, user, biome_key="TIER_1"):
        batch_size = 10
        stamina_cost = 50
        results = {
            "coins": 0,
            "dust": 0,
            "xp": 0,
            "charactersFound": [],
            "shards": 0,
            "staminaUsed": stamina_cost,
            "success": True
        }

        if user.get('stamina', 0) < stamina_cost:
            return {"success": False, "msg": f"❌ Insufficient Stamina! You need 🔋 50 Stamina for this blitz."}

        for _ in range(batch_size):
            event = self.get_random_event(1)
            if event['type'] == 'battle':
                results["coins"] += 30
                results["xp"] += 15
            else:
                reward = self.generate_step_reward(event['type'])
                results["coins"] += reward.get('coins', 0)
                results["dust"] += reward.get('dust', 0)
                results["xp"] += reward.get('xp', 0)
                results["shards"] += reward.get('shardsCurrency', 0)

        return results

    def check_daily_reset(self, user):
        if not user: return None
        now = datetime.utcnow()
        last_reset_ms = user.get('lastDailyReset', 0)
        last_reset = datetime.utcfromtimestamp(last_reset_ms / 1000.0) if last_reset_ms else datetime.min
        
        if now.day != last_reset.day or now.month != last_reset.month:
            return {
                "dailyExploreCount": 0,
                "lastDailyReset": int(now.timestamp() * 1000)
            }
        return None

explore_service = ExploreService()
